app/src/main/java/ch/epfl/sweng/erpa/services/GCP/GCPRemoteServicesProvider.java

```diff
@@ -0,0 +1,140 @@
+package ch.epfl.sweng.erpa.services.GCP;
+
+import android.support.annotation.NonNull;
+import android.support.v4.util.Consumer;
+import android.util.Log;
+
+import com.annimon.stream.Optional;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.io.IOException;
+import java.util.NoSuchElementException;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import ch.epfl.sweng.erpa.model.UserProfile;
+import ch.epfl.sweng.erpa.services.RemoteServicesProvider;
+import ch.epfl.sweng.erpa.util.GameDifficultyTypeAdapter;
+import ch.epfl.sweng.erpa.util.OptionalTypeAdapter;
+import lombok.Getter;
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+@Singleton
+public class GCPRemoteServicesProvider implements RemoteServicesProvider {
+
+    @Getter private final GCPGameService gameService;
+    @Getter private final GCPUserProfileService userProfileService;
+
+    public static final String API_URL = "https://erpa-sweng.appspot.com";
+
+    // Public for debugging purposes :)
+    public static Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(OptionalTypeAdapter.FACTORY)
+            .registerTypeAdapterFactory(GameDifficultyTypeAdapter.FACTORY)
+            .create();
+
+    @Getter @NonNull private Retrofit retrofit = new Retrofit.Builder()
+            .baseUrl(API_URL)
+            .addConverterFactory(GsonConverterFactory.create(gson))
+            .build();
+
+
+    @Inject public GCPRemoteServicesProvider() {
+        gameService = new GCPGameService(this);
+        userProfileService = new GCPUserProfileService(this);
+    }
+
+    public <T> void executeRequest(Call<T> call, Consumer<T> callback, Consumer<Exception> exceptionConsumer) {
+        Callback<T> cb0 = new Callback<T>() {
+
+            @Override public void onResponse(@NonNull Call<T> call0, Response<T> response) {
+                if (response.isSuccessful() && response.body() != null) {
+                    callback.accept(response.body());
+                } else {
+                    try {
+                        callback.accept(null);
```

Comment:
Extract the try block to a method and burry the exception (consider Exceptional?)